src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11956,6 +11991,85 @@ private ExpressionSyntax ParseNewExpression()
             }
         }
 
+        private CollectionCreationExpressionSyntax ParseCollectionCreationExpression()
+        {
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.OpenBracketToken);
+            var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
+
+            var list = _pool.AllocateSeparated<CollectionElementSyntax>();
+
+            if (this.CurrentToken.Kind != SyntaxKind.CloseBracketToken)
+            {
+tryAgain:
```

Comment:
this form is how we always do separated syntax list parsing.  I personally dont' like repeating the entire full form.  but it's what we do everywhere wlse, so this is consistent.

Reply:
Is this referring to the duplicated `IsPossibleCollectionCreationExpression() || this.CurrentToken.Kind == SyntaxKind.CommaToken` check, before and within the loop? It looks like there's also a similar `SkipBadCollectionCreationExpressionTokens()` check within and following the loop. 

If it's simpler to combine the duplicated checks in each case, within a `while (true) { ... }` for instance, I'd prefer it.

Reply:
I would very much like to keep this form in sync with all the other ParseSeparatedList algorithms we have in the parser.  if we do want to change them, doign them all at once would be my preference.

Reply:
I also really hate this code, and fwiw we are already not consistent on separated syntax list parsing algorithms. `ParseFunctionPointerType`, for example, uses a much simpler form of this.

Reply:
Funny, I was just noting in https://github.com/dotnet/roslyn/pull/65256#discussion_r1087348696 that switch expression arm parsing does it its own way too :)

I would like for `goto tryAgain;` to goto the git history and for us to do something uniformly better here. But that seems like work better done in the base branch.

Reply:
I think a very suitable approach would be to have a basic ParseSeparatedList helper which takes in a few callbacks (And some boolean options, like "allowTrailingSeparator").  Using your approach of passing the parser instance along rikki, we could ensure this doesn't cause allocs.  

But i was worried about potential perf regressions of doing that, so i wanted to leave to an alternate PR that could be tested in isolation.